README.md
# ussd
![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg?longCache=true&style=flat-square)

Run ussd code directly in your application

## Getting Started

On Android you'll need to add either the `ACCESS_CALL_PHONE` permission to your Android Manifest. To do so open the AndroidManifest.xml file (located under android/app/src/main) and add one of the following one line as direct children of the `<manifest>` tag:
``` Xml
<uses-permission android:name="android.permission.CALL_PHONE"/>
```
### main.dart

``` Dart

import 'package:flutter/material.dart';
import 'dart:async';

import 'package:ussd/ussd.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {

  @override
  void initState() {
    super.initState();
  }

  Future<void> launchUssd(String ussdCode) async {
    Ussd.runUssd(ussdCode);
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Run ussd code plugin by mandreshope'),
        ),
        body: Center(
          child: Column(
            children: <Widget>[
              RaisedButton(
                onPressed: () {
                  launchUssd("#123#");
                }, 
                child: Text("Tap to run ussd Orange Mg operator"),
              ),
              RaisedButton(
                onPressed: () {
                  launchUssd("*999#");
                }, 
                child: Text("Tap to run ussd Airtel Mg operator"),
              ),
            ],
          )
        ),
      ),
    );
  }
}

```


CHANGELOG.md
## 0.0.2+1

- Updating the documentation.

## 0.0.2

- Updating the documentation.

## 0.0.1

* Initial release.


test/ussd_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:ussd/ussd.dart';

void main() {
  const MethodChannel channel = MethodChannel('ussd');

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('runUssdCode', () async {
    expect(await Ussd.runUssd("*436*2*1*1*0330297426*0330297426#"), '42');
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:ussd_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


android/src/main/java/com/mandreshope/ussd/UssdPlugin.java
package com.mandreshope.ussd;

import android.Manifest;
import android.content.Context;
import android.app.Application;
import android.os.Bundle;
import android.os.Build;
import android.content.pm.PackageManager;
import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.widget.Toast;


import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** UssdPlugin */
public class UssdPlugin implements MethodCallHandler {
  /** Plugin registration. */
  private static final int PHONE_CALL_REQUEST_ID = 513469796;
  MethodChannel methodChannel;
  private Activity activity;
  private Registrar registrar;
  private Application.ActivityLifecycleCallbacks activityLifecycleCallbacks;
  // The code to run after requesting phone_call permissions.
  private Runnable phonecallPermissionContinuation;
  private boolean requestingPermission;
  
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "ussd");
    channel.setMethodCallHandler(new UssdPlugin(registrar, registrar.activity(), channel));
  }

  // constructor
  public UssdPlugin(Registrar registrar, Activity activity, MethodChannel methodChannel) {
    this.registrar = registrar;
    this.activity = activity;
    this.methodChannel = methodChannel;
    this.methodChannel.setMethodCallHandler(this);

    registrar.addRequestPermissionsResultListener(new PhoneCallRequestPermissionsListener());

    this.activityLifecycleCallbacks = new Application.ActivityLifecycleCallbacks() {
      @Override
      public void onActivityCreated(Activity activity, Bundle savedInstanceState) {}

      @Override
      public void onActivityStarted(Activity activity) {}

      @Override
      public void onActivityResumed(Activity activity) {
        boolean wasRequestingPermission = requestingPermission;
        if (requestingPermission) {
          requestingPermission = false;
        }
        if (activity != UssdPlugin.this.activity) {
          return;
        }
      }

      @Override
      public void onActivityPaused(Activity activity) {}

      @Override
      public void onActivityStopped(Activity activity) {}

      @Override
      public void onActivitySaveInstanceState(Activity activity, Bundle outState) {}

      @Override
      public void onActivityDestroyed(Activity activity) {}
    };
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("runUssd")) {
      String ussdCode = call.argument("ussdCode");
      runUssd(ussdCode);
    } else {
      result.notImplemented();
    }
  }

  void runUssd(String ussdCode){
    if (phonecallPermissionContinuation != null) {
      // result.error("phoneCallPermission", "Phone_Call permission request ongoing", null);
    }
    phonecallPermissionContinuation =
        new Runnable() {
          @Override
          public void run() {
            phonecallPermissionContinuation = null;
            if (!hasPhoneCallPermission()) {
              // result.error("phoneCallPermission", "MediaRecorderCamera permission not granted", null);
              return;
            }
          }
        };
    requestingPermission = false;
    if (hasPhoneCallPermission()) {
      Intent intent = new Intent(Intent.ACTION_CALL);
      intent.setData(ussdToCallableUri(ussdCode));
      try{
        activity.startActivity(intent);
      } catch (SecurityException e){
        e.printStackTrace();
      }
      phonecallPermissionContinuation.run();
    } else {
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
        requestingPermission = true;
        registrar.activity().requestPermissions(new String[] {Manifest.permission.CALL_PHONE}, PHONE_CALL_REQUEST_ID);
      }
    }
  }

  private boolean hasPhoneCallPermission() {
    return Build.VERSION.SDK_INT < Build.VERSION_CODES.M
        || activity.checkSelfPermission(Manifest.permission.CALL_PHONE)
            == PackageManager.PERMISSION_GRANTED;
  }

  private class PhoneCallRequestPermissionsListener implements PluginRegistry.RequestPermissionsResultListener {
    @Override
    public boolean onRequestPermissionsResult(int id, String[] permissions, int[] grantResults) {
      if (id == PHONE_CALL_REQUEST_ID) {
        phonecallPermissionContinuation.run();
        return true;
      }
      return false;
    }
  }


  private Uri ussdToCallableUri(String ussd) {

    String uriString = "";

    if(!ussd.startsWith("tel:"))
        uriString += "tel:";

    for(char c : ussd.toCharArray()) {

        if(c == '#')
            uriString += Uri.encode("#");
        else
            uriString += c;
    }

    return Uri.parse(uriString);
  }
   
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:ussd/ussd.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {

  @override
  void initState() {
    super.initState();
  }

  Future<void> launchUssd(String ussdCode) async {
    Ussd.runUssd(ussdCode);
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Run ussd code plugin by mandreshope'),
        ),
        body: Center(
          child: Column(
            children: <Widget>[
              RaisedButton(
                onPressed: () {
                  launchUssd("#123#");
                }, 
                child: Text("Tap to run ussd Orange Mg operator"),
              ),
              RaisedButton(
                onPressed: () {
                  launchUssd("*999#");
                }, 
                child: Text("Tap to run ussd Airtel Mg operator"),
              ),
            ],
          )
        ),
      ),
    );
  }
}
